Undefined items added to resulting mapped collection

Random `undefined` items are added to my mapped array with version 1.1.0. I don't have the same issue with version 1.0.0 or with bluebird. I've been able to consistently reproduce the issue with the following test. The same test works with concurrency 1 or concurrency >= `myInput.length`. 

```
test('async with concurrency: 2', async t => {
	const myInput = [100, 200, 10, 36, 13, 45];
	const myMapper = value => {
		return new Promise(resolve => {
			setTimeout(() => resolve(value), value);
		});
	};

	t.deepEqual(await m(myInput, myMapper, {concurrency: 2}), myInput);
});
```

It is not obvious to see the extra undefined items with the ava output (I've never used ava before) but my mocha/chai version of the test is showing this output:

```
AssertionError: expected [ 100, 200, 10, , 36, , 13, , 45 ] to deeply equal [ 100, 200, 10, 36, 13, 45 ]
      + expected - actual

       [
         100
         200
         10
      -  [undefined]
         36
      -  [undefined]
         13
      -  [undefined]
         45
       ]

```

--- a/index.js
+++ b/index.js
@@ -15,13 +15,16 @@ module.exports = (iterable, mapper, opts) => new Promise((resolve, reject) => {
 	let isRejected = false;
 	let iterableDone = false;
 	let resolvingCount = 0;
+	let currentIdx = 0;
 
-	const next = i => {
+	const next = () => {
 		if (isRejected) {
 			return;
 		}
 
 		const nextItem = iterator.next();
+		const i = currentIdx;
+		currentIdx++;
 
 		if (nextItem.done) {
 			iterableDone = true;
@@ -41,7 +44,7 @@ module.exports = (iterable, mapper, opts) => new Promise((resolve, reject) => {
 				val => {
 					ret[i] = val;
 					resolvingCount--;
-					next(i + concurrency);
+					next();
 				},
 				err => {
 					isRejected = true;
@@ -51,7 +54,7 @@ module.exports = (iterable, mapper, opts) => new Promise((resolve, reject) => {
 	};
 
 	for (let i = 0; i < concurrency; i++) {
-		next(i);
+		next();
 
 		if (iterableDone) {
 			break;

--- a/test.js
+++ b/test.js
@@ -16,13 +16,13 @@ const mapper = ([val, ms]) => delay(ms).then(() => val);
 test('main', async t => {
 	const end = timeSpan();
 	t.deepEqual(await m(input, mapper), [10, 20, 30]);
-	t.true(inRange(end(), 290, 330));
+	t.true(inRange(end(), 290, 430));
 });
 
 test('concurrency: 1', async t => {
 	const end = timeSpan();
 	t.deepEqual(await m(input, mapper, {concurrency: 1}), [10, 20, 30]);
-	t.true(inRange(end(), 590, 660));
+	t.true(inRange(end(), 590, 760));
 });
 
 test('concurrency: 4', async t => {
@@ -42,12 +42,8 @@ test('handles empty iterable', async t => {
 });
 
 test('async with concurrency: 2', async t => {
-	const myInput = [100, 200, 10, 36, 13, 45];
-	const myMapper = value => {
-		return new Promise(resolve => {
-			setTimeout(() => resolve(value), value);
-		});
-	};
-	const result = await m(myInput, myMapper, {concurrency: 2});
-	t.deepEqual(result, myInput);
+	const input = Array(10).map(() => randomInt(0, 100));
+	const mapper = value => delay(value).then(() => value);
+	const result = await m(input, mapper, {concurrency: 2});
+	t.deepEqual(result, input);
 });

--- a/test.js
+++ b/test.js
@@ -41,8 +41,15 @@ test('handles empty iterable', async t => {
 	t.deepEqual(await m([], mapper), []);
 });
 
-test('async with concurrency: 2', async t => {
+test('async with concurrency: 2 (random time sequence)', async t => {
 	const input = Array(10).map(() => randomInt(0, 100));
+	const mapper = value => delay(value).then(() => value);
+	const result = await m(input, mapper, {concurrency: 2});
+	t.deepEqual(result, input);
+});
+
+test('async with concurrency: 2 (problematic time sequence)', async t => {
+	const input = [100, 200, 10, 36, 13, 45];
 	const mapper = value => delay(value).then(() => value);
 	const result = await m(input, mapper, {concurrency: 2});
 	t.deepEqual(result, input);

--- a/test.js
+++ b/test.js
@@ -54,3 +54,10 @@ test('async with concurrency: 2 (problematic time sequence)', async t => {
 	const result = await m(input, mapper, {concurrency: 2});
 	t.deepEqual(result, input);
 });
+
+test('async with concurrency: 2 (out of order time sequence)', async t => {
+	const input = [200, 100, 50];
+	const mapper = value => delay(value).then(() => value);
+	const result = await m(input, mapper, {concurrency: 2});
+	t.deepEqual(result, input);
+});

--- a/test.js
+++ b/test.js
@@ -40,3 +40,14 @@ test('concurrency: 4', async t => {
 test('handles empty iterable', async t => {
 	t.deepEqual(await m([], mapper), []);
 });
+
+test('async with concurrency: 2', async t => {
+	const myInput = [100, 200, 10, 36, 13, 45];
+	const myMapper = value => {
+		return new Promise(resolve => {
+			setTimeout(() => resolve(value), value);
+		});
+	};
+	const result = await m(myInput, myMapper, {concurrency: 2});
+	t.deepEqual(result, myInput);
+});

